examples/all-clusters-app/esp32/main/AppTask.cpp

```diff
@@ -0,0 +1,184 @@
+    CHIP_ERROR err = CHIP_NO_ERROR;
+
+    statusLED1.Init(STATUS_LED_GPIO_NUM);
+    // Our second LED doesn't map to any physical LEDs so far, just to virtual
+    // "LED"s on devices with screens.
+    statusLED2.Init(GPIO_NUM_MAX);
+    bluetoothLED.Init();
+    wifiLED.Init();
+    pairingWindowLED.Init();
+    // Print QR Code URL
+    PrintOnboardingCodes(chip::RendezvousInformationFlags(CONFIG_RENDEZVOUS_MODE));
+
+#if CONFIG_HAVE_DISPLAY
+    InitDeviceDisplay();
+#endif
+    return err;
+}
+
+void AppTask::AppTaskMain(void * pvParameter)
+{
+    AppEvent event;
+
+    CHIP_ERROR err = sAppTask.Init();
+    if (err != CHIP_NO_ERROR)
+    {
+        ESP_LOGI(TAG, "AppTask.Init() failed due to %" CHIP_ERROR_FORMAT, err.Format());
+        return;
+    }
+
+    ESP_LOGI(TAG, "App Task started");
+
+    while (true)
+    {
+        BaseType_t eventReceived = xQueueReceive(sAppEventQueue, &event, pdMS_TO_TICKS(10));
+        while (eventReceived == pdTRUE)
+        {
+            sAppTask.DispatchEvent(&event);
+            eventReceived = xQueueReceive(sAppEventQueue, &event, 0);
+        }
+    }
+}
+
+void AppTask::PostEvent(const AppEvent * aEvent)
+{
+    if (sAppEventQueue != NULL)
+    {
+        BaseType_t status;
+        if (xPortInIsrContext())
+        {
+            BaseType_t higherPrioTaskWoken = pdFALSE;
+            status                         = xQueueSendFromISR(sAppEventQueue, aEvent, &higherPrioTaskWoken);
+        }
+        else
+        {
+            status = xQueueSend(sAppEventQueue, aEvent, 1);
+        }
+
+        if (!status)
+            ESP_LOGE(TAG, "Failed to post event to app task event queue");
+    }
+    else
+    {
+        ESP_LOGE(TAG, "Event Queue is NULL should never happen");
+    }
+}
+
+void AppTask::DispatchEvent(AppEvent * aEvent)
+{
+    if (aEvent->Handler)
+    {
+        aEvent->Handler(aEvent);
+    }
+    else
+    {
+        ESP_LOGI(TAG, "Event received with no handler. Dropping event.");
+    }
+}
+
+void AppTask::ButtonEventHandler(uint8_t btnIdx, uint8_t btnAction)
+{
+    AppEvent button_event           = {};
+    button_event.Type               = AppEvent::kEventType_Button;
+    button_event.ButtonEvent.PinNo  = btnIdx;
+    button_event.ButtonEvent.Action = btnAction;
+
+    if(btnAction==APP_BUTTON_PRESSED){
+        button_event.Handler = ButtonPressedAction;
+        sAppTask.PostEvent(&button_event);
+    }
+    
+}
+
+void AppTask::ButtonPressedAction(AppEvent * aEvent)
+{
+    vTaskDelay(50 / portTICK_PERIOD_MS);
+#if CONFIG_DEVICE_TYPE_M5STACK
+    uint32_t io_num = aEvent->ButtonEvent.PinNo;
+    int level       = gpio_get_level((gpio_num_t) io_num);
+    ESP_LOGI(TAG, "GPIO[%d] intr, val: %d, pressed\n", io_num, level);
+    if (level == 0)
+    {
+        bool woken = false;
+        if (!woken)
+        {
+            woken = WakeDisplay();
+        }
+        if (woken)
+        {
+            return;
+        }
+        ScreenManager::ButtonPressed(40 - io_num);
```

Comment:
Add a comment for what `40 - io_num` stands for.
```suggestion
        // GPIO 37 - 39 is used on M5 stack for buttons.
        ScreenManager::ButtonPressed(40 - io_num);
```